OpenDreamRuntime/Procs/DMOpcodeHandlers.cs

```diff
@@ -39,15 +39,38 @@ static class DMOpcodeHandlers {
             return null;
         }
 
-        public static ProcStatus? CreateListEnumerator(DMProcState state) {
+        public static ProcStatus? CreateListEnumerator(DMProcState state)
+        {
+            state.Pop().TryGetValueAsPath(out var type);
             DreamObject listObject = state.Pop().GetValueAsDreamObject();
             DreamList list = listObject as DreamList;
 
             if (list == null) {
                 if (listObject == null) {
                     list = null;
-                } else if (listObject.IsSubtypeOf(DreamPath.Atom) || listObject.IsSubtypeOf(DreamPath.World)) {
+                } else if (listObject.IsSubtypeOf(DreamPath.Atom))
+                {
                     list = listObject.GetVariable("contents").GetValueAsDreamList();
+                }
+                else if (listObject.IsSubtypeOf(DreamPath.World)) {
```

Comment:
`for(var/datum/D in world)` is not the same thing as `for(var/datum/D)`. This implementation leads to the two having the same behavior.

A better way to do this would be to have the compiler output a new `CreateTypeEnumerator` opcode that pushes an enumerator for the instances of a type.